libs/server/ACL/CommandPermissionSet.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Garnet.server.ACL
+{
+    /// <summary>
+    /// Wraps up command permissions behind a reference so it can be atomically swapped.
+    /// </summary>
+    public sealed class CommandPermissionSet
+    {
+        // Do not move these, initialization order is important
+        private static readonly ushort CommandListLength = GetCommandListLength();
+
+        public static readonly CommandPermissionSet All = new("+@all");
+        public static readonly CommandPermissionSet None = new("");
+
+        // Each bit corresponds to RespCommand + subcommand
+        private readonly ulong[] _commandList;
+
+        private CommandPermissionSet(string description)
+            : this(new ulong[CommandListLength], description)
+        {
+        }
+
+        private CommandPermissionSet(ulong[] commandList, string description)
+        {
+            this._commandList = commandList;
+            this.Description = description;
+
+            // Certain commands can always be run, so initialize with them
+            foreach (RespCommand command in Enum.GetValues<RespCommand>())
+            {
+                if (command.IsNoAuth())
+                {
+                    AddCommand(command);
+                }
+            }
+        }
+
+        /// <summary>
+        /// String which, when parsed by <see cref="ACLParser"/>, will produce an equivalent <see cref="CommandPermissionSet"/> to this one.
+        /// 
+        /// This is not updated automatically, and should be rationalized once modification via <see cref="RemoveCommand(RespCommand)"/> or <see cref="AddCommand(RespCommand)"/>
+        /// is complete.
+        /// </summary>
+        public string Description { get; set; }
+
+        /// <summary>
+        /// Returns true if the given command + subCommand pair can be run.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool CanRunCommand(RespCommand command)
+        {
+            // Special case "everything is permitted" 
+            if (this == All)
```

Comment:
You had the `AllState` enum earlier which seemed to be a useful concept and caused only a byte comparison here. Any reason to use this refactoring? I.e. does it help performance wise?

Reply:
Sounds good, both are fine so up to you, thanks.

Reply:
Looks like avoiding the pointer is still a bit better locally.  Preloading isn't needed or beneficial with the `IsNoAuth()` check returning, so I'll remove that.